src/awssh/autocompleter.py

```diff
@@ -0,0 +1,347 @@
+
+    Args:
+        cred_filename (Path): The path to the credential file.
+        filter (Optional[str], optional): Regular expression to match profile names. Defaults to None.
+
+    Returns:
+        set[str]: A set of profiles within a credential file that match the filter.
+    """
+    if filter:
+        filter_re: re.Pattern[str] = re.compile(filter)
+    config = configparser.ConfigParser()
+    config.read(cred_filename)
+    result: set[str] = set()
+    for section in config.sections():
+        if not filter_re or filter_re.search(section):
+            result.add(section)
+    return result
+
+
+def get_instances(
+    credential_file: Path, profile: str, region: str
+) -> set[tuple[str, str]]:
+    """Return a set of instances and names that match the profile and region.
+
+    Args:
+        credential_file (Path): The path to the credential file.
+        profile (str): The profile to use.
+        region (str): The AWS region to use.
+
+    Returns:
+        set[tuple[str, str]]: A set of instances and names that match the profile and region.
+    """
+    # Create session
+    # boto3 doesn't have a programatic way to set the credential file. yuck.
+    os.environ["AWS_SHARED_CREDENTIALS_FILE"] = str(credential_file)
+    session = boto3.session.Session(region_name=region, profile_name=profile)
+    ec2 = session.resource("ec2")
+    instances = {}
+    # Loop through instances that are running
+    for i in ec2.instances.filter(
+        Filters=[{"Name": "instance-state-name", "Values": ["running"]}]
+    ):
+        # convert tags into a proper dictionary
+        if i.tags is None:
+            # if an instance has no tags use the id as the Name
+            log(f"warning: instance does not have tags: {i}")
+            i.tag_dict = {"Name": i.id}
+        else:
+            i.tag_dict = {tag["Key"]: tag["Value"] for tag in i.tags}
+        instances[i.tag_dict["Name"]] = i
+
+    result: set[tuple[str, str]] = set()
+    for name, instance in instances.items():
+        result.add((instance.id, name))
+    return result
+
+
+def build_option_candidates(word_set: set[str]) -> set[str]:
+    """Return a set of command line option candidates for the given word set.
+
+    Args:
+        word_set (set[str]): The set of words to build option candidates for.
+
+    Returns:
+        set[str]: A set of command line option candidates for the given word set.
+    """
+    candidates: set[str] = set()
+    if HELP_OPTIONS & word_set:
+        return candidates
+    if len(word_set) <= 1:
+        candidates |= HELP_OPTIONS
+    if not CREDENTIALS_OPTIONS & word_set:
+        candidates |= CREDENTIALS_OPTIONS
+    if not LOG_LEVEL_OPTIONS & word_set:
+        candidates |= LOG_LEVEL_OPTIONS
+    if not REGION_OPTIONS & word_set:
+        candidates |= REGION_OPTIONS
+    if not SSH_ARGS_OPTIONS & word_set:
+        candidates |= SSH_ARGS_OPTIONS
+    return candidates
+
+
+class ParsedState(object):
```

Comment:
I think this would be better implemented as a [dataclass](https://docs.python.org/3/library/dataclasses.html). At minimum it should have the following change:

```suggestion
class ParsedState:
```

Reply:
Good call- please check out https://github.com/cisagov/awssh/pull/9/commits/971b187c48ae7319969c680fa79ada7caed6b3dd.